ratis-server/src/main/java/org/apache/ratis/server/storage/StorageImplUtils.java

```diff
@@ -113,61 +93,86 @@ static List<File> getExistingStorageSubs(String storageDirName, StartupOption op
-      return recoverExistingStorageDir(existingDirs.get(0), freeSpaceMin, logCorruptionPolicy);
-    } else if (option == StartupOption.FORMAT) {
-      return formatNewStorageDir(storageDirName, dirsInConf, dirsPerVol, freeSpaceMin, logCorruptionPolicy);
-    } else {
-      throw new IllegalArgumentException("Illegal option: " + option);
-    }
+    return new Op(storageDirName, option, properties).run();
   }
 
-  private static RaftStorageImpl recoverExistingStorageDir(File dir, SizeInBytes freeSpaceMin,
-      Log.CorruptionPolicy logCorruptionPolicy) throws IOException {
-    try {
-      final RaftStorageImpl storage = newRaftStorage(dir, freeSpaceMin, StartupOption.RECOVER, logCorruptionPolicy);
-      storage.initialize();
-      return storage;
-    } catch (Throwable e) {
-      if (e instanceof IOException) {
-        throw e;
+  private static class Op {
+    private final String storageDirName;
+    private final StartupOption option;
+
+    private final SizeInBytes freeSpaceMin;
+    private final Log.CorruptionPolicy logCorruptionPolicy;
+    private final List<File> dirsInConf;
+
+    private final List<File> existingSubs;
+    private final Map<File, Integer> dirsPerVol = new HashMap<>();
+
+    Op(String storageDirName, StartupOption option, RaftProperties properties) {
+      this.storageDirName = storageDirName;
+      this.option = option;
+
+      this.freeSpaceMin = RaftServerConfigKeys.storageFreeSpaceMin(properties);
+      this.logCorruptionPolicy = RaftServerConfigKeys.Log.corruptionPolicy(properties);
+      this.dirsInConf = RaftServerConfigKeys.storageDir(properties);
+
+      this.existingSubs = getExistingStorageSubs(dirsInConf, this.storageDirName, dirsPerVol);
+    }
+
+    RaftStorageImpl run() throws IOException {
+      if (option == StartupOption.FORMAT) {
+        return format();
+      } else if (option == StartupOption.RECOVER) {
+        return recover();
+      } else {
+        throw new IllegalArgumentException("Illegal option: " + option);
       }
-      throw new IOException("Failed to initialize the existing directory " + dir.getAbsolutePath(), e);
     }
-  }
 
-  private static RaftStorageImpl formatNewStorageDir(String storageDirName, List<File> dirsInConf,
-      Map<File, Integer> dirsPerVol, SizeInBytes freeSpaceMin, Log.CorruptionPolicy logCorruptionPolicy)
-      throws IOException {
-    for(List<File> volumes = new ArrayList<>(dirsInConf); !volumes.isEmpty(); ) {
-      final File dir = chooseNewStorageDir(storageDirName, dirsPerVol);
+    private RaftStorageImpl format() throws IOException {
+      if (!existingSubs.isEmpty()) {
+        throw new IOException("Failed to " + option + ": One or more existing directories found " + existingSubs
+            + " for " + storageDirName);
+      }
+
+      for (List<File> volumes = new ArrayList<>(dirsInConf); !volumes.isEmpty(); ) {
+        final File dir = chooseNewStorageDir(storageDirName, dirsPerVol);
```

Comment:
In this for loop, chooseNewStorageDir could return the same volume even if the volume has been removed in the catch block?

Reply:
@lokeshj1703 , Good catch!  The catch block should remove the vol from dirsPerVol.